controllers/front/OrderConfirmationController.php

```diff
@@ -45,31 +48,94 @@ public function init()
+        $this->id_module = (int) (Tools::getValue('id_module', 0));
+
+        // This data is kept only for backward compatibility purposes
+        $this->reference = (string) $this->order->reference;
 
+        $redirectLink = $this->context->link->getPageLink('history', $this->ssl);
+
+        // The confirmation link must contain a unique order secure key matching the key saved in database,
+        // this prevents user to view other customer's order confirmations
         if (!$this->id_order || !$this->id_module || !$this->secure_key || empty($this->secure_key)) {
             Tools::redirect($redirectLink . (Tools::isSubmit('slowvalidation') ? '&slowvalidation' : ''));
         }
-        $this->reference = $order->reference;
-        if (!Validate::isLoadedObject($order) || $order->id_customer != $this->context->customer->id || $this->secure_key != $order->secure_key) {
+
+        if (!Validate::isLoadedObject($this->order) || $this->secure_key != $this->order->secure_key) {
             Tools::redirect($redirectLink);
         }
         $module = Module::getInstanceById((int) ($this->id_module));
-        if ($order->module != $module->name) {
+        if ($this->order->module != $module->name) {
             Tools::redirect($redirectLink);
         }
-        $this->order_presenter = new OrderPresenter();
+
+        // If checks passed, initialize customer, we will need him anyway
+        $this->customer = new Customer((int) ($this->order->id_customer));
+    }
+
+    /**
+     * Logic after submitting forms
+     *
+     * @see FrontController::postProcess()
+     */
+    public function postProcess()
+    {
+        if (Tools::isSubmit('submitTransformGuestToCustomer')) {
+            // Only variable we need is the password
+            // There is no need to check other variables, because hacker would be kicked out in init(), if he tried to convert another customer
+            $password = Tools::getValue('password');
+
+            if (empty($password)) {
+                $this->errors[] = $this->trans(
+                    'To convert your account, you must enter a password.',
+                    [],
+                    'Shop.Forms.Help'
+                );
+            } elseif (strlen($password) < Validate::PASSWORD_LENGTH) {
+                $this->errors[] = $this->trans(
+                    'Your password must be at least %min% characters long.',
+                    ['%min%' => Validate::PASSWORD_LENGTH],
+                    'Shop.Forms.Help'
+                );
+            // Prevent error
+            // A) either on page refresh
+            // B) if we already transformed him in other window or through backoffice
+            } elseif ($this->customer->is_guest == 0) {
+                $this->errors[] = $this->trans(
+                    'We could not transform your guest account, it has already been transformed into a customer account. You can log in as a registered shopper.',
```

Comment:
```suggestion
                    'A customer account has already been created from this guest account. Please sign in.',
```
If I understand well, this is what you mean here? ☺️